test/test_apps/common/test_app_base.h

```diff
@@ -0,0 +1,1094 @@
+    }
+
+    static bool match(GenericFeaturesPNextNode const& requested, GenericFeaturesPNextNode const& supported) noexcept;
+
+    void combine(GenericFeaturesPNextNode const& right) noexcept;
+
+    VkStructureType sType = static_cast<VkStructureType>(0);
+    void*           pNext = nullptr;
+    VkBool32        fields[field_capacity];
+};
+
+struct GenericFeatureChain
+{
+    std::vector<GenericFeaturesPNextNode> nodes;
+
+    template <typename T>
+    void add(T const& features) noexcept
+    {
+        // If this struct is already in the list, combine it
+        for (auto& node : nodes)
+        {
+            if (static_cast<VkStructureType>(features.sType) == node.sType)
+            {
+                node.combine(features);
+                return;
+            }
+        }
+        // Otherwise append to the end
+        nodes.push_back(features);
+    }
+
+    bool match_all(GenericFeatureChain const& extension_requested) const noexcept;
+    bool find_and_match(GenericFeatureChain const& extension_requested) const noexcept;
+
+    void chain_up(VkPhysicalDeviceFeatures2& feats2) noexcept;
+
+    void combine(GenericFeatureChain const& right) noexcept;
+};
+
+GFXRECON_END_NAMESPACE(detail)
+
+enum class InstanceError
+{
+    vulkan_unavailable,
+    vulkan_version_unavailable,
+    vulkan_version_1_1_unavailable,
+    vulkan_version_1_2_unavailable,
+    failed_create_instance,
+    failed_create_debug_messenger,
+    requested_layers_not_present,
+    requested_extensions_not_present,
+    windowing_extensions_not_present,
+};
+enum class PhysicalDeviceError
+{
+    no_surface_provided,
+    failed_enumerate_physical_devices,
+    no_physical_devices_found,
+    no_suitable_device,
+};
+enum class QueueError
+{
+    present_unavailable,
+    graphics_unavailable,
+    compute_unavailable,
+    transfer_unavailable,
+    queue_index_out_of_range,
+    invalid_queue_family_index
+};
+enum class DeviceError
+{
+    failed_create_device,
+    VkPhysicalDeviceFeatures2_in_pNext_chain_while_using_add_required_extension_features,
+};
+enum class SwapchainError
+{
+    surface_handle_not_provided,
+    failed_query_surface_support_details,
+    failed_create_swapchain,
+    failed_get_swapchain_images,
+    failed_create_swapchain_image_views,
+    required_min_image_count_too_low,
+    required_usage_not_supported
+};
+
+const char* to_string_message_severity(VkDebugUtilsMessageSeverityFlagBitsEXT s);
+const char* to_string_message_type(VkDebugUtilsMessageTypeFlagsEXT s);
+
+const char* to_string(InstanceError err);
+const char* to_string(PhysicalDeviceError err);
+const char* to_string(QueueError err);
+const char* to_string(DeviceError err);
+const char* to_string(SwapchainError err);
+
+std::exception to_exception(InstanceError err);
+std::exception to_exception(PhysicalDeviceError err);
+std::exception to_exception(QueueError err);
+std::exception to_exception(DeviceError err);
+std::exception to_exception(SwapchainError err);
+
```

Comment:
All these should return `std::runtime_error`, otherwise [slicing](https://en.wikipedia.org/wiki/Object_slicing) will happen